postgres/datadog_checks/postgres/statements.py

```diff
@@ -0,0 +1,191 @@
+
+        # Given all of the available columns, determine the columns that will actually be used to query
+        desired_columns = (
+            list(PG_STAT_STATEMENTS_METRIC_COLUMNS.keys())
+            + list(PG_STAT_STATEMENTS_OPTIONAL_COLUMNS)
+            + list(PG_STAT_STATEMENTS_TAG_COLUMNS.keys())
+        )
+        self._pg_stat_statements_query_columns = sorted(list(set(desired_columns) & self._pg_stat_statements_columns))
+
+    def collect_per_statement_metrics(self, instance, db, instance_tags):
+        if not self.config.deep_database_monitoring:
+            return
+
+        try:
+            self.__collect_per_statement_metrics(db, instance_tags)
+        except Exception:
+            logger.exception('Unable to collect statement metrics due to an error')
+
+    def __collect_per_statement_metrics(self, instance, db, instance_tags):
+        self.init_pg_stat_statements_columns(db)
+        missing_columns = PG_STAT_STATEMENTS_REQUIRED_COLUMNS - self._pg_stat_statements_columns
+        if len(missing_columns) > 0:
+            logger.warning(
+                'Unable to collect statement metrics because required fields are unavailable: %s',
+                ', '.join(list(missing_columns)),
+            )
+            return
+
+        cursor = db.cursor(cursor_factory=psycopg2.extras.DictCursor)
+
+        rows = self._execute_query(
+            cursor,
+            STATEMENTS_QUERY.format(
+                cols=', '.join(self._pg_stat_statements_query_columns),
+                pg_stat_statements_view=self.config.pg_stat_statements_view,
+            ),
+        )
+        if not rows:
+            return
+
+        def row_keyfunc(row):
+            # old versions of pg_stat_statements don't have a query ID so fall back to the query string itself
+            queryid = row['queryid'] if 'queryid' in row else row['query']
+            return (queryid, row['datname'], row['rolname'])
+
+        rows = self._state.compute_derivative_rows(rows, PG_STAT_STATEMENTS_METRIC_COLUMNS.keys(), key=row_keyfunc)
+        metric_limits = (
+            self.config.statement_metric_limits if self.config.statement_metric_limits else DEFAULT_METRIC_LIMITS
+        )
+        rows = apply_row_limits(rows, metric_limits, 'calls', True, key=row_keyfunc)
+
+        for row in rows:
+            try:
+                normalized_query = datadog_agent.obfuscate_sql(row['query'])
+            except Exception as e:
+                logger.warning("Failed to obfuscate query '%s': %s", row['query'], e)
+                continue
+
+            # The APM resource hash will use the same query signature because the grouped query is close
+            # enough to the raw query that they will intersect frequently.
+            query_signature = compute_sql_signature(normalized_query)
+            apm_resource_hash = query_signature
+            tags = ['query_signature:' + query_signature, 'resource_hash:' + apm_resource_hash] + instance_tags
```

Comment:
This was to be explicit that the `query_signature` is being used as the resource hash in APM since they are two completely different concepts. Better to do that in a comment though.